server/src/main/java/org/elasticsearch/ingest/IngestCtxMap.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.ingest;
+
+import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.index.VersionType;
+import org.elasticsearch.script.CtxMap;
+import org.elasticsearch.script.Metadata;
+
+import java.time.ZonedDateTime;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Map containing ingest source and metadata.
+ *
+ * The Metadata values in {@link IngestDocument.Metadata} are validated when put in the map.
+ * _index, _id and _routing must be a String or null
+ * _version_type must be a lower case VersionType or null
+ * _version must be representable as a long without loss of precision or null
+ * _dyanmic_templates must be a map
+ * _if_seq_no must be a long or null
+ * _if_primary_term must be a long or null
+ *
+ * The map is expected to be used by processors, server code should the typed getter and setters where possible.
+ */
+class IngestCtxMap extends CtxMap {
+
+    /**
+     * Create an IngestCtxMap with the given metadata, source and default validators
+     */
+    IngestCtxMap(
+        String index,
+        String id,
+        long version,
+        String routing,
+        VersionType versionType,
+        ZonedDateTime timestamp,
+        Map<String, Object> source
+    ) {
+        super(new HashMap<>(source), new IngestMetadata(index, id, version, routing, versionType, timestamp));
+    }
+
+    /**
+     * Create IngestCtxMap from a source and metadata
+     *
+     * @param source the source document map
+     * @param metadata the metadata map
+     */
+    IngestCtxMap(Map<String, Object> source, Metadata metadata) {
+        super(source, metadata);
+    }
+
+    /**
+     * Fetch the timestamp from the ingestMetadata, if it exists
+     * @return the timestamp for the document or null
+     */
+    public static ZonedDateTime getTimestamp(Map<String, Object> ingestMetadata) {
+        if (ingestMetadata == null) {
+            return null;
+        }
+        Object ts = ingestMetadata.get(IngestDocument.TIMESTAMP);
+        if (ts instanceof ZonedDateTime timestamp) {
+            return timestamp;
+        } else if (ts instanceof String str) {
+            return ZonedDateTime.parse(str);
+        }
+        return null;
+    }
+
+    static class IngestMetadata extends Metadata {
```

Comment:
Since this is static, let's move it out to a tope level class. It can still be package private.

Reply:
There is already an [IngestMetadata](https://github.com/elastic/elasticsearch/blob/f8f440af95c68e8934e2472b7472f77fcf1a1923/server/src/main/java/org/elasticsearch/ingest/IngestMetadata.java) that "Holds the ingest pipelines that are available in the cluster".

I'm going to rename this to `IngestDocMetadata` when I make it top-level.